### PR TITLE
[HACKATHON] - Implement project search filter

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://trello.com/c/9IJg1qR8/40-add-a-search-bar-to-search-for-project-name-id-in-the-projects-list

#### Links to any Pull Requests related to this

N.A.

#### Anything else the reviewers should know about the changes here

- Search by project name, project id or experiment id.

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by vickymorrison** [comment]
The full text in the search bar is not visibile - I suggest shortening the instruction and/or adding a tooltip. Other than that, this new feature is great!

Screenshot 2021-09-03 at 14 04 48

File: src/__test__/components/data-management/ProjectCard.test.jsx
Comment (cosa65) [inline src/__test__/components/data-management/ProjectCard.test.jsx:38]: While we're at it we could add some more tests based on variation of the props

File: src/components/data-management/ProjectCard.jsx
Comment (cosa65) [inline src/components/data-management/ProjectCard.jsx]: You could extract the const {Item} = Descriptions, right? Might make this a little bit more readable

File: src/components/data-management/ProjectsListContainer.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsListContainer.jsx]: Remove this

File: src/components/data-management/ProjectsListContainer.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsListContainer.jsx]: I feel like these shouldn't be constants that are recalculated on every render, I would move them to functions or even better, if they are only used in one useEffect then inside those

File: src/components/data-management/ProjectsListContainer.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsListContainer.jsx]: Move this inside `createNewProject`

File: src/components/data-management/ProjectsListContainer.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsListContainer.jsx]: After moving the other stuff you can check if this can be moved too

File: src/components/data-management/ProjectsList.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsList.jsx:13]: Is there some way we can find out the `height` without receiving it from outside?

File: src/components/data-management/ProjectsListContainer.jsx
Comment (marcellp) [reply src/components/data-management/ProjectsListContainer.jsx]: I believe @ivababukova has fixes that clears these up.

File: src/components/data-management/ProjectsListContainer.jsx
Comment (marcellp) [reply src/components/data-management/ProjectsListContainer.jsx]: I believe @ivababukova has fixes that clears these up.

File: src/components/data-management/ProjectsListContainer.jsx
Comment (marcellp) [reply src/components/data-management/ProjectsListContainer.jsx]: I believe @ivababukova has fixes that clears these up.

**Review by marcellp — APPROVED** [review]

**Review by marcellp — DISMISSED** [review]

**Review by marcellp — DISMISSED** [review]
I think this should wait until https://github.com/biomage-ltd/ui/pull/494 by @ivababukova is approved and merged. You will potentially waste a lot of time on feedback on code that is actually being fixed up in that other PR and you probably will need to do some conflict resolution after that's merged in.

Other than that, the code looks good.

**Comment by cosa65** [comment]
> I think this should wait until #494 by @ivababukova is approved and merged. You will potentially waste a lot of time on feedback on code that is actually being fixed up in that other PR and you probably will need to do some conflict resolution after that's merged in.
> 
> Other than that, the code looks good.

Good point, @aerlaut let me kno when you need a new code review for this PR

File: src/__test__/components/data-management/ProjectCard.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/ProjectCard.test.jsx:38]: Hmm.. I don't think there's much else to check. I added checks for the remaining attributes to be shown.

File: src/components/data-management/ProjectCard.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectCard.jsx]: Done 👍

File: src/components/data-management/ProjectsListContainer.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsListContainer.jsx]: no more

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx:13]: I had a quick search on this matter. It seems that we may be able to use a `ref` that points to an HTML element that's rendered, and get the `ref.current.offsetWidth` value to get the width. I tried this on the elements rendered by this function, but it returns undefined. According to [this comment](https://stackoverflow.com/a/64814269/1940886) on StackOverflow, using the above method we may be able to get the initial width of the object. However, to listen we still need to set up a resize listener. We currently do this using the `ReactResizeObserver` HOC component, which is the source of the `height` props seen here.

File: src/components/data-management/ProjectsListContainer.jsx
Comment (cosa65) [inline src/components/data-management/ProjectsListContainer.jsx]: what's 'ig'?

Comment (aerlaut) [reply src/components/data-management/ProjectsListContainer.jsx]: It's the flags for Regex : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp

i : case insensitive search
g : global search - find all matches instead of stopping at the first one

**Review by cosa65 — APPROVED** [review]

File: src/components/data-management/ProjectsListContainer.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsListContainer.jsx]: I think the `g` flag can be removed as we only need the first hit, not every occurence of the string. It'll help speed up the searching. I'll change this.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: Here you send different variations of `project.uuid` as props, which is not great because of multiple reasons:
- Creates unnecessary coupling between `ProjectsList` and `ProjectCard`, now `ProjectsList` has to know some internal information about how exactly and what `ProjectCard` needs in order to compute stuff.
- Makes `ProjectCard` component harder to use in other places -- for any other place that needs to use `ProjectCard`, that place will need to work out and pass over all these things.
- Code is hard to read and hard to maintain.

Very, very visible example is the `validateInputs` function: so that you use it in `ProjectCard`, you have to now import it in `ProjectsList`. Why? It will only make `ProjectsList` become our next `ProjectDetails` monster.

You can easily remove all these props and have that handled internally from within `ProjectCard`.

File: src/components/data-management/ProjectCard.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectCard.jsx:36]: Can you settle on one convention and make sure that all component names follow it? Are we going to use Project**s** for all component names that are about multiple projects? Then rename this and every other component to be `Projects<blablabal>`. If not, then again make sure all components are consistent.

File: src/components/data-management/ProjectCard.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectCard.jsx]: It will make the code cleaner I feel if you also have a variable for the style of the inactive project and use it here similarly to `activeProjectStyle`:

```
const inactiveProjectStyle = {
    cursor: 'pointer'
}

....

style={isActive ? activeProjectStyle : inactiveProjectStyle}
```

File: src/__test__/components/data-management/ProjectCard.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/ProjectCard.test.jsx:37]: after you move the functionality of `ProjectCard` to be within the component itself, you will need to move the unit tests from `ProjectsList` to here.

File: src/__test__/components/data-management/ProjectsList.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/ProjectsList.test.jsx:91]: I don't see any unit tests to test for the functionality of the `matcher` function.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: Rename this -- throughout the code, in the tests we already use this name `matcher` and it is also a package. I would call the function something like `filterByRegex`. Also, what is this `searchObj`? Is this the list of all projects? I think we need rename that too, it is confusing.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: same here as for the tons of props you are passing in `ProjectCard`  -- you you really need to have these as props to `ProjectDeleteModal`?

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: what's the point of this function `searchable`? I am really confused here.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: you have 1 loop inside this `searchable` function, now you have another loop. Is that necessary? Can you try to avoid adding all those loops?

File: src/components/data-management/ProjectCard.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectCard.jsx:36]: I was differentiating it depending on whether the component relates to a single project (e.g. **Project**Card) or multiple projects (e.g. **Projects**List). However, in the interest of clarity, I think it's better to use **Projects** for the prefix of all components, to have it in line with the other parts too.

Comment (ivababukova) [reply src/components/data-management/ProjectCard.jsx:36]: Ok, so there was a reasoning behind that differentiating, good, this is all I was asking for. Then I would leave it to `ProjectCard` if it applies to a single project

Comment (aerlaut) [reply src/components/data-management/ProjectCard.jsx:36]: On 2nd thought, I think the current naming is good. This is called `ProjectCard` because it contains only 1 project, not multiple project.

`ProjectsList` is a list of projects
`ProjectsListContainer` is the container for `ProjectsList`.

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: This is reduced to a variable that stores the value of comparisons instead of a function

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: It was made to make it easier to search. It's removed in latest refactor.

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: Yes, refactored.

File: src/__test__/components/data-management/ProjectsList.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/ProjectsList.test.jsx:91]: It's the test on line 110

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx:35]: this `key` is not being used in the `ProjectCard` component -- I think you forgot to delete it. Also, make sure you add a key to each Card rendered within ProjectCard, because right now when I open the code locally, I can see a warning message about `Warning: Each child in a list should have a unique "key" prop.`

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: rename this to `ProjectsList` -- the components names have to match the file names.

Comment (ivababukova) [reply src/components/data-management/ProjectsList.jsx]: Just noticed that we have this `ProjectsList` component defined both under `ProjectsListContainer.jsx` and `ProjectsList.jsx` files. This must be wrong, make sure you delete whichever of them is not used.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx:47]: This is not right -- I just checked index.jsx page, you are not passing `filter` in there. Either correct index.jsx, or mark this prop as not required and make sure the code won't fail if it is missing.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectsList.jsx]: I just checked what happens if we don't define height and guess -- nothing. The component looks the same. So you can remove all these height passings happening in this component and also from the other `ProjectsListContainer` component from from index.jsx when you call it and this should remove some extra unneeded code.

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [reply src/components/data-management/ProjectsList.jsx:47]: Actually I worked it out, there are two components called `ProjectsListContainer`, which you will separately need to fix.

File: src/components/data-management/ProjectCard.jsx
Comment (ivababukova) [inline src/components/data-management/ProjectCard.jsx]: instead of saying `labelStyle={{ fontWeight: 'bold' }}` could you not use styled components: https://styled-components.com/ and create a new `BoldItem` component (at the top of this file since it is internal) and use it here instead of `Item`?

File: src/__test__/components/data-management/ProjectsList.test.jsx
Comment (ivababukova) [inline src/__test__/components/data-management/ProjectsList.test.jsx:145]: sounds great, but we also need some more tests on what will happen if: we there was nothing matched (so there are 0 results) and what if we pass in none instead of filter, also case sensitivity.

Comment (ivababukova) [reply src/__test__/components/data-management/ProjectsList.test.jsx:145]: there might be more cases too -- have you seen if the coverage report shows whether there are any of the lines not covered by unit tests and which ones?

**Review by ivababukova — CHANGES_REQUESTED** [review]
way better, but there is still some small things to fix

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: Refactored

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: Refactored into `ProjectCard`

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx:35]: The error is actually because of line 25 in `ProjectsListContainer`. The `Skeleton`s doesn't have keys.  I've added them in fbbbaa4

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: Whoops, my bad. fixed

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: The height here is added to constraint the height of the Space element and force it to show a scroll bar. If we remove height, when you add more projects, the scrollbar will not appear as the element's height expands to accommodate the projects, so you can not scroll to select the projects below.

Comment (ivababukova) [reply src/components/data-management/ProjectsList.jsx]: Ok, you are right, we can't remove it. But then, why `height: height - 90`? Why not `height: height - 10`? Or `-5`?

File: src/components/data-management/ProjectsList.jsx
Comment (ivababukova) [reply src/components/data-management/ProjectsList.jsx:35]: great!

File: src/components/data-management/ProjectCard.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectCard.jsx]: I've tried using Styled components but the styles aren't transferred to the Item texts. So, I refactored the style out instead.

File: src/components/data-management/ProjectsList.jsx
Comment (aerlaut) [reply src/components/data-management/ProjectsList.jsx]: This is addressed in commit dd8faee. The value is changed to 130. The value comes from adding padding, button size and gap between buttons in the UI. The breakdown can be seen in `ProjectsList` line 8-16

File: src/__test__/components/data-management/ProjectsList.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/ProjectsList.test.jsx:145]: Additional cases added, including checking for searches with `projectUuid` and `experimentId`

File: src/components/data-management/ProjectCard.jsx
Comment (ivababukova) [reply src/components/data-management/ProjectCard.jsx]: ok, sounds good

**Review by ivababukova — APPROVED** [review]

**Pull request merged**
